vsintegration/src/FSharp.Editor/Formatting/EditorFormattingService.fs

```diff
@@ -0,0 +1,101 @@
+﻿// Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+namespace Microsoft.VisualStudio.FSharp.Editor
+
+open System.Composition
+open System.Collections.Generic
+
+open Microsoft.CodeAnalysis
+open Microsoft.CodeAnalysis.Editor
+open Microsoft.CodeAnalysis.Host.Mef
+open Microsoft.CodeAnalysis.Text
+
+open Microsoft.FSharp.Compiler.SourceCodeServices
+open System.Threading
+
+[<Shared>]
+[<ExportLanguageService(typeof<IEditorFormattingService>, FSharpConstants.FSharpLanguageName)>]
+type internal FSharpEditorFormattingService
+    [<ImportingConstructor>]
+    (
+        checkerProvider: FSharpCheckerProvider,
+        projectInfoManager: ProjectInfoManager
+    ) =
+
+    static member GetFormattingChanges(document: Document, sourceText: SourceText, checker: FSharpChecker, optionsOpt: FSharpProjectOptions option, position: int) =
+        // Logic should be:
+        // If first token on the current line is a closing brace,
+        // match the indent with the indent on the line that opened it
+
+        asyncMaybe {
+            let! options = optionsOpt
+            
+            let line = sourceText.Lines.[sourceText.Lines.IndexOf position]
```

Comment:
IndexOf throws if the position isn't within the text, but as Roslyn has already supplied us with the position I think it's impossible that we can't find the line in the source text (also supplied by Roslyn).